frontend/components/dynamic-badge-maker.js

```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+function escapeField(s) {
+  return encodeURIComponent(s.replace(/-/g, '--').replace(/_/g, '__'));
+}
+
+export default class DynamicBadgeMaker extends React.Component {
+  makeBadgeUri () {
+    let url = this.props.baseUri;
+    url += '/dynamic/' + escapeField(this.typeInput.value);
+    url += '.svg?label=' + escapeField(this.labelInput.value);
+    url += '&colorB=' + escapeField(this.colorInput.value);
+    url += '&prefix=' + escapeField(this.prefixInput.value);
+    url += '&suffix=' + escapeField(this.suffixInput.value);
+    url += '&query=' + this.queryInput.value;
+    url += '&uri=' + escapeField(this.uriInput.value);
+    return url;
+  }
+
+  handleSubmit (e) {
+    e.preventDefault();
+    document.location = this.makeBadgeUri();
+  }
+
+  render() {
+    return (
+      <form onSubmit={e => this.handleSubmit(e)}>
+        <input
+          className="short"
+          ref={input => { this.typeInput = input; }}
+          name="type"
+          list="dynamic-type"
+          placeholder="type" />
+        <datalist id="dynamic-type">
+          <option value="json" />
+        </datalist>
+        <input
+          className="short"
+          ref={input => { this.labelInput = input; }}
```

Comment:
This is not really the "React way" of doing things. React components should rarely directly touch the raw DOM elements. Instead of doing this, you should use controlled components (https://reactjs.org/docs/forms.html)

For example:
```
<input
  className="short"
  value={this.state.label}
  onChange={evt => this.setState({label: evt.target.value})
/>
```